src/bin/workload_generator.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Anza Technology, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::net::{SocketAddr, UdpSocket};
+use std::time::{Duration, Instant};
+
+use alpenglow::crypto::aggsig;
+use alpenglow::crypto::signature::SecretKey;
+use alpenglow::network::NetworkMessage;
+use alpenglow::{Transaction, ValidatorInfo, logging};
+use clap::Parser;
+use color_eyre::Result;
+use log::info;
+use rand::RngCore;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct ConfigFile {
+    id: u64,
+    identity_key: SecretKey,
+    #[serde(deserialize_with = "aggsig::SecretKey::from_array_of_bytes")]
+    voting_key: aggsig::SecretKey,
+    port: u16,
+    gossip: Vec<ValidatorInfo>,
+}
+
+/// Worklaod generator for benchmarks.
+#[derive(Debug, Clone, Parser)]
+#[command(version, about, long_about = None)]
+struct Args {
+    /// Validator addres or receiving transactions.
+    #[arg(long)]
+    validator: String,
+    /// Initial wait time, before sending any transactions.
+    #[arg(long)]
+    initial_delay_secs: Option<u64>,
+    /// Target throughput in transactions per second.
+    #[arg(long)]
+    transactions_per_second: Option<u64>,
+    /// Bytes per transaction.
+    #[arg(long)]
+    transaction_size: Option<u64>,
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    // enable fancy `color_eyre` error messages
+    color_eyre::install()?;
+
+    logging::enable_logforth();
+
+    // parse args & load config from file
+    let args = Args::parse();
+
+    let validator_addr = args.validator;
+    let addr: SocketAddr = validator_addr.parse().unwrap();
+    let target_tps = args.transactions_per_second.unwrap_or(100);
+    let bytes_per_tx = args.transaction_size.unwrap_or(512);
+
+    // let network = UdpNetwork::new_with_any_port();
+    let socket = UdpSocket::bind("0.0.0.0:0")?;
+
+    tokio::time::sleep(Duration::from_secs(args.initial_delay_secs.unwrap_or(0))).await;
+
+    let start_time = Instant::now();
+    let mut last_report = Instant::now();
+    let mut txs_sent = 0;
+
+    let mut rng = rand::rng();
+    let mut buf = vec![0; bytes_per_tx as usize];
+    socket.set_nonblocking(true).unwrap();
```

Comment:
I would generally expect a udp socket to never block but doesn't hurt to have this.  I think we could add a comment like the following though:

// making nonblocking so if the generator is not keeping up, it fails instead of artificially slowing down

Reply:
I think we can even explicitly ignore errors when sending, because we don't really care if any packets are dropped, as long as we count the number of messages leaving the generator correctly